app/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,228 @@
+package tech.relaycorp.letro.awala
+
+import android.content.Context
+import android.util.Log
+import androidx.annotation.RawRes
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+import tech.relaycorp.awaladroid.Awala
+import tech.relaycorp.awaladroid.GatewayBindingException
+import tech.relaycorp.awaladroid.GatewayClient
+import tech.relaycorp.awaladroid.endpoint.FirstPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.InvalidThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.PublicThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.ThirdPartyEndpoint
+import tech.relaycorp.awaladroid.messaging.OutgoingMessage
+import tech.relaycorp.letro.R
+import tech.relaycorp.letro.awala.message.Message
+import tech.relaycorp.letro.awala.message.MessageRecipient
+import tech.relaycorp.letro.awala.message.MessageType
+import tech.relaycorp.letro.ui.navigation.Route
+import tech.relaycorp.letro.utils.awala.loadNonNullFirstPartyEndpoint
+import tech.relaycorp.letro.utils.awala.loadNonNullThirdPartyEndpoint
+import javax.inject.Inject
+
+interface AwalaManager {
+    val messages: Flow<Message>
+    suspend fun sendMessage(
+        message: Message,
+        recipient: MessageRecipient,
+    )
+    suspend fun isAwalaInstalled(currentScreen: Route): Boolean
+}
+
+class AwalaManagerImpl @Inject constructor(
+    private val awalaRepository: AwalaRepository,
+    @ApplicationContext private val context: Context,
+): AwalaManager {
+
+    private val awalaScope = CoroutineScope(Dispatchers.IO)
+
+    private val _messages = MutableSharedFlow<Message>()
+    override val messages: Flow<Message>
+        get() = _messages
+
+    @Volatile
+    private var isAwalaSetUp = false
+    private var awalaSetupJob: Job? = null
+
+    @Volatile
+    private var isAwalaInstalledOnDevice: Boolean? = null
+
+    @Volatile
+    private var isReceivingMessages = false
+
+    init {
+        awalaSetupJob = awalaScope.launch {
+            Awala.setUp(context)
+            checkIfAwalaAppInstalled()
+            isAwalaSetUp = true
+            awalaSetupJob = null
+        }
+    }
+
+    override suspend fun sendMessage(
+        message: Message,
+        recipient: MessageRecipient,
+    ) {
```

Comment:
I think that this method -- in its current form -- is too greedy and complex (it does a lot more than just send messages), and it will only get more complex as we continue to develop the app. This is a version of the method that I certainly see:

```kotlin
suspend fun sendMessage(message: Message, recipient: ThirdPartyEndpoint) {
    GatewayClient.sendMessage(
        OutgoingMessage.build(
            type = message.type.value,
            content = message.content,
            senderEndpoint = firstPartyEndpoint,
            recipientEndpoint = recipient,
        )
    )
}
```

`firstPartyEndpoint` will be used a lot, so I think we should just cache it in the class instance or use dependency injection.

As for the third-party endpoint, I think it should be retrieved by the caller and any failure to create/retrieve it should be local to that caller. Namely:

- **The Letro server's third-party endpoint**: Like the first-party endpoint, this one will be used a lot, so I think we should cache it in the instance of this class or use dependency injection. We could even create a method somewhere called `sendMessageTo[Letro]Server()`, which would use this method behind the scenes.
- **Contact's third-party endpoint**: For example, this could be a method on the `Contact` instance (or whatever we call that class) to retrieve the third-party endpoint (e.g., `contact.getEndpoint()`).
- **The third-party endpoint to claim non-Relaycorp-managed accounts** (e.g., "alice@bbc.com"). For example, this could be a method on the account (e.g., `account.getVeraidAuthorityEndpoint()`).

Reply:
Good comment, thanks for making it such detailed! 

I think that resolving of endpoints is responsibility of the AwalaManager, since it's Awala-specific logic. More than that, now this resolving is performed in one place. 

But I agree that this method is kind of big, so I'm going to decompose it by moving the logic of resolving endpoints to private methods

What do you think?

Reply:
> I think that resolving of endpoints is responsibility of the AwalaManager, since it's Awala-specific logic. More than that, now this resolving is performed in one place.

Let's forget about Awala for a moment. If you were building an HTTP client inside the Android app and it communicated with a RESTful API that offered many endpoints across various modules (e.g., account creation, contact pairing, messaging), would you implement the processing of every request and every response in the same class?

Reply:
FYI, I'm mainly thinking of this in the context of the final set of features we'll need from it, not just the functionality we need today. So maybe the best approach here is to keep the `GatewayManager` as it is but create a task to refactor it later, once we can see everything it does. Wdyt?

Reply:
If we're using HTTP-terms, I consider AwalaManager like HTTPClient, which has abstract methods for sending HTTP-requests to a server. So yes, I would implement one HTTPClient, and it's a responsibility of callers to pass to this client endpoint and body of a request

Reply:
But you said that `GatewayManager` had some bugs or I'm confusing something?

Reply:
Nope, not that I know of. My concerns are purely to do with maintainability and easy of understanding... Especially because I can see myself working a lot around this area (even though I won't go anywhere close to the UI).

But like I said in my previous comment, I think we should move on and just keep an eye on it as the app evolves. We can defer any further refactoring until then.

Reply:
Let me clarify, you want to proceed with `GatewayRepository` or `AwalaManager`? Because we used the term `GatewayManager` which doesn't exist :)

Reply:
If  it's `GatewayRepository`, I don't see how this is possible to fast switch the implementation from `AwalaManager` to  `GatewayRepository`, and don't see any point to do it neither.

Reply:
(unless the `AwalaManager` has some critical issues in current realisation aside from architecture of `sendMessage()` method)

Reply:
I meant `AwalaGateway`, sorry. I've just updated the comment above too.

Again, I'm saying we should leave it as it is, especially after the changes you made, but keep an eye on it as we implement more functionality as a refactoring may be needed.

Reply:
Sure, now it's clear! I just got scaried that you proposed to use `GatewayRepository` as it is :)

Reply:
Haha, sorry for the confusion! 😄